Feat/upsell banners hfg

preda-bogdan opened the pull request:
### Summary
<!-- Please describe the changes you made. -->
Created a new Control and Section that will display an Upsell banner inside the Customizer.
These banners are used with the existing Upsell class to add the banners to Layout Single Post, Layout Archive, Header, and Footer.

To not introduce new strings inside the theme I reused the main upsell text for the Footer banner.

Once a banner is dismissed it will dismiss all banner instances, this was intended on my part as I think it would be a bad user experience to close multiple banners individually.

### Will affect the visual aspect of the product
<!-- It includes visual changes? -->
YES

### Screenshots <!-- if applicable -->
Single Post (pic. 1)
![image](https://user-images.githubusercontent.com/23024731/236176777-5485e2a6-3284-48b3-84f5-d280d182c887.png)

Archive (pic. 2)
![image](https://user-images.githubusercontent.com/23024731/236177177-4c896745-08c0-4723-8877-61bc26af4d2f.png)

Header (pic. 3)
![image](https://user-images.githubusercontent.com/23024731/236177394-8c81d59d-c8f2-4487-9beb-7b3d0009a701.png)

Footer (pic. 4)
![image](https://user-images.githubusercontent.com/23024731/236177498-66f89b34-10e7-49b6-865f-36b91317df25.png)



### Test instructions
<!-- Describe how this pull request can be tested. -->
1. Use only Neve without PRO being active.
2. Go to Customizer to **Layout** > **Single Post** and check the upsell banner at the bottom of the page. See pic. 1
3. Go to Customizer to **Layout** > **Blog / Archive** and check the upsell banner at the bottom of the page. See pic. 2
4. Go to Customizer to **Header** and check the upsell banner. See pic. 3
5. Go to Customizer to **Footer** and check the upsell banner. See pic. 4
6. Check that when dismissed the banner will not appear anymore. The expiry time is set to 7 days. You can use Transient Manager to remove the transient and check that the banner is displayed again. The transient name is `upsell_dismiss_banner_customizer`

## Check before Pull Request is ready:

* [ ] I have [written a test](CONTRIBUTING.md#writing-an-acceptance-test) and included it in this PR
* [x] I have [run all tests](CONTRIBUTING.md#run-tests) and they pass
* [x] The code passes when [running the PHP CodeSniffer](CONTRIBUTING.md#run-php-codesniffer)
* [x] Code meets [WordPress Coding Standards](CONTRIBUTING.md#coding-standards) for PHP, HTML, CSS and JS
* [x] [Security and Sanitization](CONTRIBUTING.md#security-and-sanitization) requirements have been followed
* [x] I have assigned a reviewer or two to review this PR (if you're not sure who to assign, we can do this step for you)

<!-- Issues that this pull request closes. -->
Closes #3964.
<!-- Should look like this: `Closes #1, #2, #3.` . -->

abaicus reviewed (commented):
@preda-bogdan code looks good to me! 🚀 I only have one question:

I can see that we dismiss all upsells in the backend, but they will still appear in the current customizer session, e.g.:
- go to footer
- dismiss upsell
- go to header
- upsell is still there

I would find this a bit annoying seeing as I already dismissed it once in the customizer session, and the backend actually hides all of them for the next 7 days. 

If we want to dismiss all of them, I suggest adding a global state and pass it as a prop [here](https://github.com/Codeinwp/neve/pull/3972/files#diff-f9963a0ae9a5b87cd59b3c75f05f00809264a29bda0575901a6c9482c464561eR246) to mark all banners as dismissed. 

Let me know if this is the intended behavior! 🚀 Thank you!

preda-bogdan commented:
@abaicus Thank you for mentioning this, I made some changes to the dismissal. However, I could not use a global state as the components are mounted independently and are not part of a React App. I did use a custom event to sync all states, this seems to do the job as well.

Let me know if this solution is good from your point of view, and if you have other observations.
I will await your feedback before proceeding to QA.

abaicus commented on assets/apps/customizer-controls/src/builder-upsell/BannerUpsell.tsx:
```diff
@@ -2,11 +2,22 @@ import React, { useState } from 'react';
 
 type Props = {
 	control: WPCustomizeControl;
+	useHideBannerState: () => {
+		hide: boolean;
+		setHide: (hide: boolean) => void;
+	};
```
I think we can remove this, otherwise everything looks good 👍🏻

irinelenache commented:
@preda-bogdan Nice job 🚀 Tested the upsells and here is what i found:

- [ ] There is an empty space in the **Blog/Archive** panel, left from the previous upsell https://vertis.d.pr/i/3eOmi9
- [ ] This is how the **header upsell** looks on tablet https://vertis.d.pr/i/hI1tSD
- [ ] Adding a component in header/footer when the upsell is visible will look like this https://vertis.d.pr/i/VP6tj3
- [ ] Is it ok that the footer won't be visible when the **footer upsell** is present? https://vertis.d.pr/i/RH6T3O

preda-bogdan commented:
@irinelenache Thank you for testing this and for your findings. 💯  I addressed the issues you mentioned. For the last one that references the footer not being visible it should not be an issue, as the user can hide banner if he is not interested. I don't see it as an issue. 

Let me know if everything works as expected now. 🚀

irinelenache commented:
@preda-bogdan Tested the fixes and everything is fine now, thank you 🚀
